homeassistant/components/supla/__init__.py

```diff
@@ -107,26 +150,42 @@ def discover_devices(hass, hass_config):
                 continue
 
             channel["server_name"] = server_name
-            component_configs.setdefault(component_name, []).append(channel)
+            component_configs.setdefault(component_name, []).append(
+                {
+                    "channel_id": channel_id,
+                    "server_name": server_name,
+                    "function_name": channel["function"]["name"],
+                }
+            )
 
     # Load discovered devices
-    for component_name, channel in component_configs.items():
-        load_platform(hass, component_name, "supla", channel, hass_config)
+    for component_name, config in component_configs.items():
+        await async_load_platform(hass, component_name, "supla", config, hass_config)
 
 
 class SuplaChannel(Entity):
     """Base class of a Supla Channel (an equivalent of HA's Entity)."""
 
-    def __init__(self, channel_data):
-        """Channel data -- raw channel information from PySupla."""
-        self.server_name = channel_data["server_name"]
-        self.channel_data = channel_data
+    def __init__(self, config):
```

Comment:
I suggest directly passing coordinator via parameter as itll be easily to work with if this becomes a config entry later.

Reply:
```suggestion
    def __init__(self, config, coordinator):
```

Reply:
Hmm, I tried that. Passing coordinator directly gives me a recorder warning similar to the one below:

```
WARNING (Recorder) [homeassistant.components.recorder]
Event is not JSON serializable:
```

Is that OK? I also prefer to pass the coordinator and server directly, but I didn't want to introduce new warnings.

Reply:
that seems odd, i know many integrations that do it this way. lets do it via pass in and look into the warning as it might be something specific this integration does

Reply:
OK, looking at rest of your proposed changes I get it now. The warning was a result of passing server and coordinator in the configuration dictionary. Thanks!